### ISSUE TITLE
[HOLD for payment 2023-03-14] [$1000] Report item remains hovered even when mouse is moved away from it.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15300]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Hover over a message other than the first one in a group of message in the chat history (i.e any message that does not have the author avatar alongside it).
2. Move the cursor to the mini context menu (top right corner).
3. Right click on the mini context menu.
4. Close the big context menu by clicking anywhere outside the highlighted row.

## Expected result:
When clicked outside the item to close the main context menu, the mini context menu should disappear along with the hovered row.

## Actual result:
Both the hovered row and mini context menu remain visible.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Chrome / Safari

**Version Number:** 1.2.74-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/220150346-a37cce86-4cdb-4cf1-b674-0a7d79d0aeaf.mov

https://user-images.githubusercontent.com/43996225/220150354-df3936ba-354d-4fc9-a9af-1957765144b9.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @allroundexperts 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676595049049579

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bff08adf85d6111c</li>
        <li>Upwork Job ID: 1628009672196374528</li>
        <li>Last Price Increase: 2023-03-16</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by trjExpensify** [comment]
Okay, I can repro this reliably: 
![aR6mNKNuzs](https://user-images.githubusercontent.com/16232057/220343503-630497d0-1830-40e1-8551-c63a9c4e6d4d.gif)

Took me a bit to figure out the steps, so I'm updating the OP to stress it's any message that isn't the first in a group of messages.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01bff08adf85d6111c

**Comment by MelvinBot** [comment]
Current assignee @trjExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @aimane-chnaif (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @yuwenmemon (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Both the hovered row and mini context menu remain visible when we hover over the mini context menu and right click to open the main context menu, then click outside.

### What is the root cause of that problem?
The root cause is when we right click, the `isContextMenuActive` will become false and then the  `MiniReportActionContextMenu` will be hidden immediately, before the `onMouseLeave` of the `Hoverable` is triggered.

Due to this the `Hoverable` is stuck in the `hovered: true` state, even though the cursor already moves away from it.

### What changes do you think we should make in order to solve the problem?

To fix this we can simply remove this line https://github.com/Expensify/App/blob/df3c6e94787d796818def86a8b8d211a77a7ee38/src/pages/home/report/ReportActionItem.js#L168

When the main context menu is opened, the `onMouseLeave` of `Hoverable` will be triggered and `hovered` will become `false`, then `MiniReportActionContextMenu` will hide, and the row will lose it's hovered state as expected.

### What alternative solutions did you explore? (Optional)
NA

**Comment by tienifr** [comment]
@getusha I think `resetHoverStateOnOutsideClick` was intentionally removed because it's a hack (see [this comment](https://github.com/Expensify/App/issues/14080#issuecomment-1379459244)).

Just curious if you find any other cases besides this one that requires `resetHoverStateOnOutsideClick`? Because it seems for this case the root cause [is known](https://github.com/Expensify/App/issues/15300#issuecomment-1438471427) and can be fixed without reviving the  `resetHoverStateOnOutsideClick` workaround.

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Report action item remains hovered after opening main context menu inside mini context menu

### What is the root cause of that problem?
I partly agree with @tienifr and @getusha. This is indeed related to the removal of `resetHoverStateOnOutsideClick` which was used to handle such cases. This is another case where `mouseLeave` event is missing or it's effect being negated. I didn't go much deeper into this issue due to the nature of my suggested solution.

Apart from the reported issue, there is another issue which is the visual flickering that occurs after right click.

### What changes do you think we should make in order to solve the problem?
Do not open main context menu inside mini context menu. The easiest way to achieve this is by re-structuring the components [here](https://github.com/Expensify/App/blob/fc6ed80290360e202520b4fdf4f72b35e5dca1bf/src/pages/home/report/ReportActionItem.js#L207).

Current structure:
```js
<PressableWithSecondaryInteraction>
    <Hoverable>
        <View>
            <UnreadActionIndicator />
            <View />
            <MiniReportActionContextMenu />
        </View>
    </Hoverable>
    <View/>
</PressableWithSecondaryInteraction>
```

Suggested structure:
```js
<>
    <Hoverable>
        <View>
            <UnreadActionIndicator />
            <PressableWithSecondaryInteraction>
                <View />
             </PressableWithSecondaryInteraction>
            <MiniReportActionContextMenu />
        </View>
    </Hoverable>
    <View/>
</>
```

### What alternative solutions did you explore? (Optional)
- Prevent showing main context menu inside mini context menu by listening to `contextmenu` event on the mini context menu and stop its propagation.

**Comment by s77rt** [comment]
I have taken the time to investigate this further and I can confirm that @tienifr RCA is correct and the suggested solution is good as well. The `mouseLeave` event is not firing because the target of that even which is the mini context menu no longer exists.

@tienifr I think you have a tiny typo here "when we right click, the isContextMenuActive will become false" I believe you meant to write true not false. Other than that, good work! :+1:

**Comment by tienifr** [comment]
> I think you have a tiny typo here "when we right click, the isContextMenuActive will become false" I believe you meant to write true not false

@s77rt yes, thanks for noticing that!

**Comment by bernhardoj** [comment]
@tienifr proposal will make a regression where the mini context menu does not hide when we right click the action item until we move the cursor (Safari only). I'm pretty sure it's because of the `onMouseLeave` Safari issue that both of you should remember from the tooltip issue [here](https://github.com/Expensify/App/issues/13146) @aimane-chnaif @s77rt.


https://user-images.githubusercontent.com/50919443/221363118-0737aa46-9436-49ae-bc93-0a9a8155c6f6.mov

**Comment by s77rt** [comment]
@bernhardoj Nice catch! The rule is usually if you fix one thing and broke another then you fixed nothing but not sure if this applies here since it only effect Safari (Safari only emits `mouseLeave` on actual mouse movements).

**Comment by s77rt** [comment]
@aimane-chnaif There are two props that have been removed: `resetsOnClickOutside` and `resetHoverStateOnOutsideClick`.
- The first one was removed because it caused a [bug](https://github.com/Expensify/App/issues/13146) and it's not the nature behaviour to dismiss a tooltip on click.
- The second one was removed because it's a hack and relaying on blur event seemed as a good replacement.

Just a small clarification, the listed PRs are not directly related to each other, for instance https://github.com/Expensify/App/pull/14377 has no relation with https://github.com/Expensify/App/pull/14048 (i.e. not related to the removal of `resetsOnClickOutside`) it just happens that the corresponding bugs were reported in a close timeline and I was involved in both 😁.

As for this issue, I think we should go with @tienifr 's [proposal](https://github.com/Expensify/App/issues/15300#issuecomment-1438471427) and do nothing for Safari since the bug is browser related and very minor.

**Comment by allroundexperts** [comment]
I agree with @s77rt here. I also think that we shouldn't block a fix because of faulty Safari behaviour.

**Comment by s77rt** [comment]
@aimane-chnaif Ah sorry forgot to comment about your concern (option 2)

https://github.com/Expensify/App/issues/14080 is not very easy to reproduce, I believe it would still exist if we revert the PR. But I think by reverting the PR you only mean to bring back `resetHoverStateOnOutsideClick`. Because that PR did two things:
- Set hover status to false on blur (This is required to fix the bug)
- Removed `resetHoverStateOnOutsideClick` (This has been removed only because it's no longer necessary, we can bring it back but let's not do it)

**Comment by s77rt** [comment]
@getusha `resetHoverStateOnOutsideClick` used to create a `document` even listener (`mousedown`) for every rendered hoverable element and those are a lot. If you single click, dozens of callbacks will be executed.

**Comment by s77rt** [comment]
@aimane-chnaif I also think that https://github.com/Expensify/App/pull/14507 has no relation with `resetHoverStateOnOutsideClick`.

**Comment by s77rt** [comment]
> yes, but I linked just because it was regression from https://github.com/Expensify/App/pull/14377 and related to `onBlur`

Correct but `onBlur` is also unrelated to this issue.

> yes, so Option 2 is just to bring back `resetHoverStateOnOutsideClick`

Yes, that's right :+1:

**Comment by tienifr** [comment]
> @tienifr do you have an updated proposal considering this safari issue?

@aimane-chnaif I agree there's a Safari bug when the View goes behind a Modal but there's no mouse movement, the onMouseLeave is not called.

If we want to handle this case, we can add specifically tell the `Hoverable` that it's behind a modal so it can reset the hovered state correctly. We can add a `isBehindModal={this.state.isContextMenuActive}` flag, and use the following logic in `componentDidUpdate` of `Hoverable`:

```
componentDidUpdate(prevProps) {
        if (!prevProps.isBehindModal && this.props.isBehindModal) {
            this.setIsHovered(false);
        }
    }
```

I've tested and it works well, I think this additional fix is also quite lightweight and doesn't require listening to any browser event.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by roryabraham** [comment]
> It seems relying on only event listeners (onMouseEnter, onMouseLeave, onBlur) without resetHoverStateOnOutsideClick causes many problems on safari ... @roryabraham what do you think? (since you suggested to remove resetHoverStateOnOutsideClick https://github.com/Expensify/App/issues/14080#issuecomment-1379459244)

I stand by this – `resetHoverStateOnOutsideClick` was a hack. If there are bugs in Safari desktop such that it doesn't correctly emit standard browser events, we should treat those like any other upstream bug and not accept workarounds but rather find a way to polyfill or fix those bugs upstream.

**Comment by yuwenmemon** [comment]
@aimane-chnaif Sounds good to me! Let's go with @tienifr's proposal.

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @yuwenmemon!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by tienifr** [comment]
Hi @aimane-chnaif @yuwenmemon, [the PR](https://github.com/Expensify/App/pull/15476) is ready for review. Please help me check it. Thank you!

**Comment by tienifr** [comment]
@yuwenmemon It would be great if you can check the PR when you are available!

**Comment by tienifr** [comment]
@yuwenmemon @aimane-chnaif 

# Regression Test Proposal

**Bug**: Screen is flashing when changing focus between edit input

**Proposed Test Steps**:

1. Hover over a message other than the first one in a group of messages in the chat history (i.e any message that does not have the author avatar alongside it).
2. Move the cursor to the mini context menu (top right corner).
3. Right-click on the mini context menu.
4. Close the big context menu by clicking anywhere outside the highlighted row.
5. Verify that the above mini context menu is disappeared

Do we :+1: or :-1:

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.79-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15476


If no regressions arise, payment will be issued on 2023-03-14. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@aimane-chnaif / @yuwenmemon] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@aimane-chnaif / @yuwenmemon] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@aimane-chnaif / @yuwenmemon] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@trjExpensify] Determine if we should create a regression test for the bug.
- [ ] [@aimane-chnaif] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to the [appropriate location](https://github.com/Expensify/App/blob/main/contributingGuides/TestRail_Navigation_C%2B.md) to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Review the proposed regression test steps and location.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by trjExpensify** [comment]
Okay, how are we looking with these items? 

>[@aimane-chnaif / @yuwenmemon] The PR that introduced the bug has been identified. Link to the PR:
> [@aimane-chnaif / @yuwenmemon] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
>[@aimane-chnaif / @yuwenmemon] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:


To prepare for payments. It looks like melvin increased the price while the proposal was under review, so I'm going to change the title back to $1k. To confirm the amounts due: 

- @tienifr: $1,000 + $500 #urgency bonus 
- @aimane-chnaif: $1,000 + $500 #urgency bonus 
- @allroundexperts: $250 for the bug report 

Is that correct?

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by trjExpensify** [comment]
Yeah, I tend to agree that this is sort of edge case to include in the PR checklist.

**Comment by trjExpensify** [comment]
On a similar note, I don't think we'd add a regression test for this. It's the sort of visual bug that would be caught by executing other existing test cases. Do you guys agree with that?

**Comment by trjExpensify** [comment]
Okay, @aimane-chnaif, confirming these are correct right? 

https://github.com/Expensify/App/issues/15300#issuecomment-1472585081

**Comment by trjExpensify** [comment]
Settled up with @aimane-chnaif!

**Comment by trjExpensify** [comment]
Settled up with @allroundexperts. Over to you, @tienifr and then we can close this out. 👍

**Comment by trjExpensify** [comment]
@tienifr settled up! closing this out, thanks everyone!
